### ISSUE TITLE
Grad-Function for which only vmapped-version produces NaN-values

### ISSUE THREAD

**Issue by simon-bachhuber** [jax-ml/jax#9159]
I will admit i am not confident this is in fact a bug. Still it seems like odd behaviour and it took me a few hours to pin-point the NaN-source.

Minimal reproducing example: 
```python
import jax.numpy as jnp 
import jax 

def safe_norm(x):
    """Grad-safe for x=0.0
    """
    assert x.ndim == 1

    is_zero = jnp.all(jnp.isclose(x, 0.0), axis=-1, keepdims=False)
    return jax.lax.cond(is_zero, 
        lambda x: jnp.array([0.0]), 
        lambda x: jnp.linalg.norm(x, keepdims=True), 
        x)

def safe_normalize(x):
    """Execution- and Grad-safe for x=0.0
    """
    assert x.ndim == 1
    
    is_zero = jnp.allclose(x, 0.0)
    return jax.lax.cond(is_zero,
        lambda x: jnp.zeros_like(x),
        lambda x: x/safe_norm(x),
        x)

@jax.grad
def output(x):
    assert x.ndim == 1
    return jnp.sum(safe_normalize(x))
```

```python
# **Crucial**! If `False` then output is as expected in both cases
jax.config.update("jax_debug_nans", True)
output(jnp.zeros((3,)))
DeviceArray([0., 0., 0.], dtype=float32)
```

```python
jax.vmap(output)(jnp.zeros((4,3,)))
FloatingPointError
```

So, my 2 cents are that when you vmap-the function he somehow starts to un-roll the two possibilities and hence hits an intermediate NaN which raises an error. If nan-debugging is disabled, then the intermediate NaN is still there it just never makes it out ;)

Either way this behaviour makes it difficult to debug for other NaN-sources because it renders the NaN-debugging useless.

**Comment by jakevdp** [comment]
It looks like you're running into this issue: https://jax.readthedocs.io/en/latest/faq.html#gradients-contain-nan-where-using-where

You can fix it using the double-where (or double-cond) trick mentioned there.

**Comment by SimiPixel** [comment]
Works, thank you! 

Out of curiosity, does this mean that the gradient is calculated using reverse-mode autodiff only in case of the vmapped-grad-function?

**Comment by jakevdp** [comment]
I don't know why it's different for the vmapped case.

**Comment by jakevdp** [comment]
Thanks for following up! Since there hasn't been any other activity here in a few years, I'll assume it's safe to close.

**Comment by alik-git** [comment]
Hi there, I would like to request / suggest that this caveat be added to this page in the docs.

https://github.com/jax-ml/jax/blob/b74b16f9b9fd13bd90dc9ed8683c54660913de0a/docs/debugging/print_breakpoint.md?plain=1#L272

Currently, there is nothing on that page to warn the user that the provided conditional breakpoint example simply won't work if used within a vmap. I believe this would save many users (and myself from the past) a bunch of time.
